test_runner/src/main/kotlin/ftl/reports/outcome/TestOutcome.kt

```diff
@@ -0,0 +1,41 @@
+package ftl.reports.outcome
+
+import com.google.api.services.toolresults.model.Environment
+import com.google.api.services.toolresults.model.Outcome
+import com.google.api.services.toolresults.model.Step
+import ftl.json.getDetails
+import ftl.util.StepOutcome
+
+data class TestOutcome(
+    val outcome: String,
+    val matrixId: String,
+    val testDetails: String
+)
+
+fun List<Environment>.createMatrixOutcomeSummaryUsingEnvironments(
+    testMatrixId: String,
+    outcome: Outcome? = getOutcomeFromEnvironments(),
+    testDetails: String? = outcome?.getDetails(map { it.createTestSuitOverviewData() }.foldTestSuiteOverviewData())
+) = TestOutcome(
+    outcome = outcome?.summary ?: "Unknown",
+    matrixId = testMatrixId,
+    testDetails = testDetails ?: "Unknown outcome"
+)
+
+private fun List<Environment>.getOutcomeFromEnvironments(): Outcome? = maxBy {
+    StepOutcome.order.indexOf(it.environmentResult?.outcome?.summary)
+}?.environmentResult?.outcome
+
+fun List<Step>.createMatrixOutcomeSummaryUsingSteps(
```

Comment:
> This is a completely subjective opinion and of course, I might be wrong (please, keep it in mind). I am not entirely convinced with this way of implementing functions. `createMatrixOutcomeSummaryUsingSteps` has 3 parameters but actually only first is used (passed) at all. I can see this could be rewritten to something like
> 
> ```kotlin
> fun List<Step>.createMatrixOutcomeSummaryUsingSteps(testMatrixId: String): TestOutcome {
>     val (summary, testDetails) = getOutcomeFromSteps().let { it?.summary to it?.getDetails(createTestSuitOverviewData()) }
>     return TestOutcome(
>         outcome = summary ?: "Unknown",
>         matrixId = testMatrixId,
>         testDetails = testDetails ?: "Unknown outcome"
>     )
> }
> ```
> 
> And looks pretty the same, but IMO it is more readable and is not confusing with 3-arg signature.
> 
> [EDIT]
> I came up with something even shorter but it is not readable at all 😺
> 
> ```kotlin
> fun List<Step>.createMatrixOutcomeSummaryUsingSteps2(testMatrixId: String) = this
>     .run { getOutcomeFromSteps().let { it?.summary to it?.getDetails(createTestSuitOverviewData()) } }
>     .run { TestOutcome(outcome = first ?: "Unknown", matrixId = testMatrixId, testDetails = second ?: "Unknown Outcome") }
> ```

it could be even
```
getOutcomeFromSteps().let { it?.summary to it?.getDetails(createTestSuitOverviewData()) }
    .run { ((summary, testDetails)) -> TestOutcome(outcome = summary ?: "Unknown", matrixId = testMatrixId, testDetails = testDetails ?: "Unknown Outcome") 
```

Reply:
It's great that you are asking if you have some concerns. 
So just tell me which version looks simpler and more clean for you?
```kotlin
fun List<Step>.createMatrixOutcomeSummaryUsingSteps2(testMatrixId: String) = this
    .run { getOutcomeFromSteps().let { it?.summary to it?.getDetails(createTestSuitOverviewData()) } }
    .run { TestOutcome(outcome = first ?: "Unknown", matrixId = testMatrixId, testDetails = second ?: "Unknown Outcome") }
```
or
```kotlin
fun List<Step>.createMatrixOutcomeSummaryUsingSteps(
    testMatrixId: String,
    outcome: Outcome? = getOutcomeFromSteps(),
    testDetails: String? = outcome?.getDetails(createTestSuiteOverviewData())
) = TestOutcome(
    outcome = outcome?.summary ?: "Unknown",
    matrixId = testMatrixId,
    testDetails = testDetails ?: "Unknown outcome"
)
```
Think about what advantages your solution brings? I only see some additional noise that comes for words like `this`, `run`, `let`, `it`, `first`, and `second`.